dialects/linux/dproc.c

```diff
@@ -1518,12 +1522,90 @@ process_proc_map(p, s, ss)
 		efs = sv = 1;
 	    else
 		efs = 0;
-	    if (!efs) {
-		if (HasNFS)
-		    sv = statsafely(fp[6], &sb);
-		else
-		    sv = stat(fp[6], &sb);
+
+	    /* kernel version >= 3.8.0 and the target process has a different
+	     * mount namespace with this lsof process. */
+	    if ((compare_kernel_release("3.8.0") >= 0) &&
+		compare_namespace("mnt")) {
+		/* Fork a child process and create a pipe. */
+                if (pipe(pipefd) < 0) {
+                    fprintf(stderr, "%s: can't open pipes: %s\n",
+                        Pn, strerror(errno));
+                    goto stat_directly;
+                }
+
+                pid = fork();
+                if (pid < 0) {
+                    fprintf(stderr, "%s: can't fork: %s\n",
+                        Pn, strerror(errno));
+                    goto stat_directly;
+                }
+
+                if (pid == 0) {
+	            /* Child process:
+	             * Move this child process into the mount namespace of the
+	             * target process before applying stat(2).
+	             */
+                    int nsfd;
+                    char nspath[100];
+
+                    close(pipefd[0]);
+                    sprintf(nspath, "/proc/%d/ns/mnt", Lp->pid);
+                    nsfd = open(nspath, O_RDONLY);
+
+                    if (nsfd < 0 || setns(nsfd, CLONE_NEWNS)) {
+                        fprintf(stderr, "%s: can't setns: %s\n",
+                            Pn, strerror(errno));
+			Exit(LSOF_ERROR);
+                    }
+
+                    if (!efs) {
+                        if (HasNFS)
+                            sv = statsafely(fp[6], &sb);
+                        else
+                            sv = stat(fp[6], &sb);
```

Comment:
Should we add some flag to tell the user that the file actually lies in another mount namespace? It can be confusing for the user to see that a process opens a file but the user cannot find that file in the default namespace.

Reply:
I considered add a flag, e.g. 'in another mnt namespace' and append it to nmabuf. Since current code on nmabuf is a little redundant, I prefer to do it in another patch focused on nmabuf handling.